### PR TITLE
Refactoring midpresentation (issue #61, #28)

### LINKED ISSUE #28: 'Adding Papers to My Collections' Pop-up
To bridge the gap between papers and collections, we need to implement pop-up where users can add papers to their own collections. The appearance of it would look like this.
[Add Papers to Collections Pop-up(2_4)](https://github.com/swsnu/swpp2019-team3/wiki/Requirements-and-Specification-Document#2_4-save-1)

### LINKED ISSUE #61: Transfer midpresentation-branch To master
In recent days, we hastily developed our service in `midpresentation-branch` for mid presentation demo. As a result, there are a lot of codes which need to be modified and refactored and some dummy data should be deleted. Also, many features lack their tests and there are some issues like #59 handled in the appropriate way, not by ad-hoc codes. So we need to split and improve changes in the `midpresentation-branch`, and gradually merge them to `master`.

### PR DESCRIPTION
**Pull request by kosumin**
Related Issue: #61 #28
It resolves #61 and resolves #28.

# Major changes
No big changes from `midpresentation-branch`. However, this PR has quite many changes.
- Adding collection reducer and collection actions
- Connecting AddPaperModal, CollectionList, CollectionDetail with actions and reducers
- Preparing connection between ProfileDetail and related actions and reducers.

Originally these changes were implemented by @njsh4261, @davin111, and me.

# Minor changes

1. json.loads is deleted from test_put_paper_colelction.
2. All test errors are fixed.

### Checklist
This pull request...
- [x] has preceding issues related with it (if resolves some issues, metions them with 'resolve')
- [ ] doesn't include too many changes (you should focus on the specific feature in a PR)
- [x] elaborately explains what feature is added, which part is fixed or improved
- [x] has step by step instructions for collaborators, if needed
- [x] has attached screenshots, if appropriate

My codes...
- [ ] include enough tests for added and changed parts
- [x] passe all the new and existing tests
- [x] include appropriate comments
- [x] don't introduce unnecessary warnings

### CODE REVIEW

File: frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js [hunk 1]
```diff
@@ -1,26 +1,81 @@
...
+            // eslint-disable-next-line react/no-unused-state
+            paperCount: 0,
             newReplyContent: "",
-            replies: this.props.thisCollection.replies,
-            isLiked: this.props.thisCollection.isLiked,
+            isLiked: false,
+            // members: [],
+            replies: [],
+            papers: [{
+                source: "recently added",
```
Comment (davin111) [inline frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js]: How about deleting these dummy data? Now, we have related actions and reducers!

Comment (kosumin) [reply frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js]: I delete this!

File: frontend/src/store/reducers/UserReducer.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (davin111) [inline frontend/src/store/reducers/UserReducer.js]: For now, I think this file shouldn't be included in this PR.

File: frontend/src/store/actions/Review/ReviewActions.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable max-len */
```
Comment (davin111) [inline frontend/src/store/actions/Review/ReviewActions.js]: For now, I think this file shouldn't be included in this PR.

File: frontend/src/store/actions/Profile/ProfileActions.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/* eslint-disable max-len */
```
Comment (davin111) [inline frontend/src/store/actions/Profile/ProfileActions.js]: For now, I think this file shouldn't be included in this PR.

File: frontend/src/index.js [hunk 5]
```diff
@@ -12,7 +12,7 @@ import store, { history } from "./store/store";
 
 ReactDOM.render(
     <Provider store={store}>
-        <App history={history} />
+        <App history={history} me={store.getState().auth.me} />
```
Comment (davin111) [inline frontend/src/index.js]: Oh, you use a different way to get 'me' from the way I use in #71 PR. Could you check the PR and let me know what you think better?

File: frontend/src/containers/Review/ReviewDetail/ReviewDetail.js [hunk 6]
```diff
@@ -117,7 +123,7 @@ class ReviewDetail extends Component {
                                     ) : null}
                                 </div>
                                 <Form className="new-reply">
-                                    <Form.Label className="username">Username </Form.Label>
+                                    <Form.Label className="username">Girin </Form.Label>
```
Comment (davin111) [inline frontend/src/containers/Review/ReviewDetail/ReviewDetail.js]: I think it would be better that we choose the master version here.

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js [hunk 7]
```diff
@@ -23,10 +24,14 @@ class AddPaperModal extends Component {
         this.openAddPaperHandler = this.openAddPaperHandler.bind(this);
         this.checkHandler = this.checkHandler.bind(this);
         this.clickCancelButtonHandler = this.clickCancelButtonHandler.bind(this);
+        this.clickAddButtonHandler = this.clickAddButtonHandler.bind(this);
     }
 
     componentDidMount() {
-
+        this.props.onGetCollections({ id: 1 })
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: I think there's no auth.me.id at the current master version.

File: frontend/src/components/Header/Header.js [hunk 8]
```diff
@@ -44,8 +44,8 @@ class Header extends Component {
                         <Dropdown>
                             <Dropdown.Toggle title="myaccount" className="myaccount-button">My Account</Dropdown.Toggle>
                             <Dropdown.Menu className="myaccount-menu">
-                                <Dropdown.Header>User className</Dropdown.Header>
-                                <Dropdown.Item className="my-profile-button" href="/profile/user_id">My Profile</Dropdown.Item>
+                                <Dropdown.Header>Girin</Dropdown.Header>
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Header/Header.js]: I think it's no matter what is written here, since it would be changed after the current user information will be delivered.

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js [hunk 9]
```diff
@@ -56,6 +61,33 @@ class AddPaperModal extends Component {
         });
     }
 
+    clickAddButtonHandler() {
+        if (this.state.collectionName !== "") {
+            this.props.onMakeNewCollection({ title: this.state.collectionName, text: " " })
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: What would be good? e.g This is '{ collectionName }' collection?

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js [hunk 10]
```diff
@@ -56,6 +61,33 @@ class AddPaperModal extends Component {
...
+                .then(() => {
+                    this.setState({
+                        makeNewCollectionStatus: collectionStatus.SUCCESS,
+                    });
+                    this.props.onAddPaper({
+                        id: this.props.id,
+                        collection_ids: [this.props.selectedCollection.id],
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: I tried to do it using checkedCollectins. However, the second 'if' is executed before making a new collection, namely before the first 'if' is finished.

File: frontend/src/components/Header/Header.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Header/Header.js]: I think it is no matter what is here, since it should be changed to the real user name after getMeStatus is added.

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: There's no action to get user information in this branch. It could be done after more merging

File: frontend/src/containers/Collection/CollectionList/CollectionList.js [hunk 11]
```diff
@@ -1,54 +1,58 @@
...
     constructor(props) {
         super(props);
         this.state = {
+            // makeNewCollectionStatus: collectionStatus.NONE,
+            collections: [],
         };
     }
 
+    componentDidMount() {
+        this.props.onGetCollections({ id: 1 })
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/containers/Collection/CollectionList/CollectionList.js]: There's no action to get user information in this branch. It could be done after more merging

File: frontend/src/containers/Collection/CollectionList/CollectionList.js [hunk 12]
```diff
@@ -59,62 +63,34 @@ class CollectionList extends Component {
...
-        source: PropTypes.string,
-        id: PropTypes.number,
-        user: PropTypes.string,
-        title: PropTypes.string,
-        paperCount: PropTypes.number,
-        replyCount: PropTypes.number,
-        likeCount: PropTypes.number,
-        isLiked: PropTypes.bool,
-    })),
+    // me: PropTypes.objectOf(PropTypes.any),
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/containers/Collection/CollectionList/CollectionList.js]: There's no action to get user information in this branch. It could be done after more merging

File: frontend/src/containers/PaperDetail/PaperDetail.test.js [hunk 13]
```diff
@@ -107,7 +225,36 @@ describe("<PaperDetail />", () => {
                 selectedPaper: {},
             },
             auth: {},
-            collection: {},
+            collection: {
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/containers/PaperDetail/PaperDetail.test.js]: I added it, becasue PaperDetail imports PaperSpec and it has addPaper Modal which requires collection state.

File: frontend/src/containers/Review/ReviewDetail/ReviewDetail.js [hunk 14]
```diff
@@ -107,7 +113,7 @@ class ReviewDetail extends Component {
                             <div className="reply">
                                 <div className="review-extra">
                                     <Button className="like-button" variant="light" onClick={this.state.isLiked ? this.clickUnlikeButtonHandler : this.clickLikeButtonHandler}><div className="heart-image"><SVG name="heart" height="70%" width="70%" /></div>{this.state.likeCount}</Button>
-                                    <Button className="replyCount-button" variant="light"><div className="reply-image"><SVG name="zoom" height="60%" width="60%" /></div>{this.state.replyCount}</Button>
+                                    <Button className="replyCount-button" variant="light"><div className="reply-image"><SVG name="zoom" hegiht="60%" width="60%" /></div>{this.state.replyCount}</Button>
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/containers/Review/ReviewDetail/ReviewDetail.js]: I will fix it again! Thank you!

File: frontend/src/containers/Review/ReviewDetail/ReviewDetail.js (hunk 6 above)
Comment (kosumin) [reply frontend/src/containers/Review/ReviewDetail/ReviewDetail.js]: ReviewDetail and ReviewControl would be changed after merging backend api for reviews. So now, it is not important whether it is Grin or Usrename(both just text), since it would be modified to {userName} soon.

File: frontend/src/index.js (hunk 5 above)
Comment (kosumin) [reply frontend/src/index.js]: I will delete it. Thank you!

File: frontend/src/store/actions/Profile/ProfileActions.js (hunk 4 above)
Comment (kosumin) [reply frontend/src/store/actions/Profile/ProfileActions.js]: I will delete this! Thank you!

File: frontend/src/store/actions/Review/ReviewActions.js (hunk 3 above)
Comment (kosumin) [reply frontend/src/store/actions/Review/ReviewActions.js]: I will delete this file. Thanks

File: frontend/src/store/reducers/UserReducer.js (hunk 2 above)
Comment (kosumin) [reply frontend/src/store/reducers/UserReducer.js]: Thank you!

File: frontend/src/containers/PaperDetail/PaperDetail.css [hunk 15]
```diff
@@ -1,50 +0,0 @@
-.paperdetail {
```
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/containers/PaperDetail/PaperDetail.css]: Oh, I will make .css file again. Sorry!

File: frontend/src/components/Header/Header.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/components/Header/Header.js]: You’re right. Then will the codes be added another PR by another person?

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: OK

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: Aha, then tentatively, I think it would be ok.

File: frontend/src/components/Header/Header.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Header/Header.js]: In fact, this problem is already fixed on your pr. So I think after merging two branches, it would be solved!

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: Oh, it’s good, I think.

File: frontend/src/containers/PaperDetail/PaperDetail.test.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/containers/PaperDetail/PaperDetail.test.js]: Now I see, sorry.

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: clickAddButtonHandler() {
        let collectionIds = this.state.checkedCollections;
        const paperId = this.props.id;
        let collectionsAndPaper = { id: paperId, collection_ids: collectionIds };
        if (this.state.collectionName.length > 0) {
            const newCollection = { title: this.state.collectionName, text: "empty description" };
            // console.log(newCollection);
            this.props.onMakeNewCollection(newCollection)
                .then(() => {
                    switch (this.props.makeNewCollectionStatus) {
                    case collectionStatus.SUCCESS:
                        this.setState({
                            addPaperCollectionStatus: collectionStatus.NONE,
                            makeNewCollectionStatus: collectionStatus.SUCCESS,
                            isAddPaperOpen: false,
                            checkedCollections: [],
                            collectionName: "",
                        });
                        collectionIds = collectionIds.concat(this.props.createdCollection.id);
                        collectionsAndPaper = { id: paperId, collection_ids: collectionIds };
                        this.props.onAddPaper(collectionsAndPaper);
                        break;
                    default:
                        this.setState({ makeNewCollectionStatus: collectionStatus.FAILURE });
                        break;
                    }
                });
        } else if (this.state.checkedCollections) {
            // console.log(collectionsAndPaper);
            this.props.onAddPaper(collectionsAndPaper)
                .then(() => {
                    switch (this.props.addPaperCollectionStatus) {
                    case collectionStatus.SUCCESS:
                        this.setState({
                            addPaperCollectionStatus: collectionStatus.SUCCESS,
                            makeNewCollectionStatus: collectionActions.NONE,
                            isAddPaperOpen: false,
                            checkedCollections: [],
                            collectionName: "",
                        });
                        break;
                    default:
                        this.setState({ addPaperCollectionStatus: collectionStatus.FAILURE });
                        break;
                    }
                });
        }
    }

[the remark this answers is not in the document]
Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: Would it be better to handle this problem like this?

File: frontend/src/components/Header/Header.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/components/Header/Header.js]: Got it.

**Comment by kosumin** [comment]
> Actually, I planed to modify AddPaperModal and related components, but you did most works! I thought you would work only about reducers, CollectionList, and CollectionDetail. Thank you very much!
> 
> By the way, I worry about many test warnings this PR generates. If you want help with tackling them(even though it is not difficult), just let me know.

I checked this problem and solved it!

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (davin111) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: Yeah, I think it is better.

File: frontend/src/components/Modal/AddPaperModal/AddPaperModal.js [hunk 16]
```diff
@@ -23,10 +27,23 @@ class AddPaperModal extends Component {
...
         this.clickCancelButtonHandler = this.clickCancelButtonHandler.bind(this);
+        this.clickAddButtonHandler = this.clickAddButtonHandler.bind(this);
     }
 
     componentDidMount() {
+        this._isMounted = true;
 
+        this.props.onGetCollections({ id: 1 })
+            .then(() => {
+                if (this._isMounted) {
```
Comment (davin111) [inline frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: Why does this code check `_isMounted`?

Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: It is for handling test error.

Comment (kosumin) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: https://www.robinwieruch.de/react-warning-cant-call-setstate-on-an-unmounted-component

Comment (davin111) [reply frontend/src/components/Modal/AddPaperModal/AddPaperModal.js]: Thanks for letting me know.

**Comment by davin111** [comment]
I think now we can merge this PR after @njsh4261 reviews.

File: frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js [hunk 17]
```diff
@@ -101,31 +137,32 @@ class CollectionDetail extends Component {
...
                             </div>
                             <div id="memberStat">
-                                <h5 id="memberCount">{this.props.thisCollection.members.length}</h5>
+                                <h5 id="memberCount">{this.state.userCount}</h5>
                                 <h5 id="memberText">Members</h5>
                             </div>
                             <div id="collectionButtons">
                                 {likeOrUnlike}
                                 {inviteButton}
-                                {this.props.thisCollection.amIMember ? editButton : <div />}
```
Comment (njsh4261) [inline frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js]: Edit Button on Collection Detail page should not be displayed if a user is not a member of this collection or an owner of it. Is it a reason for the change that there is no user related Redux files?

Comment (njsh4261) [reply frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js]: If this change is made by a mistake, please commentize the original line for now.

Comment (kosumin) [reply frontend/src/containers/Collection/CollectionDetail/CollectionDetail.js]: Sorry. I thought it would not be required, since there's no features about shared collection. I added it again.
